File: personalizatio-sdk/src/main/java/com/personalizatio/stories/models/Story.java
package com.personalizatio.stories.models;

import androidx.annotation.NonNull;

import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

final public class Story {
	private final int id;
	private final String avatar;
	private final String name;
	private boolean viewed;
	private final boolean pinned;
	private int startPosition;
	private final List<Slide> slides;

	public Story(@NonNull JSONObject json) {
		id = json.optInt("id", 0);
		avatar = json.optString("avatar", "");
		name = json.optString("name", "");
		viewed = json.optBoolean("viewed", false);
		pinned = json.optBoolean("pinned", false);
		startPosition = json.optInt("start_position", 0);
		slides = new ArrayList<>();
		var slidesJsonArray = json.optJSONArray("slides");
		if (slidesJsonArray != null) {
			for( int i = 0; i < slidesJsonArray.length(); i++ ) {
				slides.add(new Slide(slidesJsonArray.optJSONObject(i)));
			}
		}
	}

	public int getId() {
		return id;
	}

	public String getAvatar() {
		return avatar;
	}

	public String getName() {
		return name;
	}

	public boolean isViewed() {
		return viewed;
	}

	public void setViewed(boolean viewed) {
		this.viewed = viewed;
	}

	public boolean isPinned() {
		return pinned;
	}

	public int getStartPosition() {
		return startPosition;
	}

	public void setStartPosition(int startPosition) {
		this.startPosition = startPosition;
	}

	public Slide getSlide(int position) {
		return slides.get(position);
	}

	public int getSlidesCount() {
		return slides.size();
	}

	public void resetStartPosition() {
		if (startPosition >= getSlidesCount() || startPosition < 0 ) {
			startPosition = 0;
		}
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Story story)) return false;
		return id == story.id
				&& viewed == story.viewed
				&& pinned == story.pinned
				&& startPosition == story.startPosition
				&& avatar.equals(story.avatar)
				&& name.equals(story.name)
				&& slides.equals(story.slides);
	}

	@Override
	public int hashCode() {
		return Objects.hash(id, avatar, name, viewed, pinned, startPosition, slides);
	}

	@NonNull
	@Override
	public String toString() {
		return String.format("Story{id='%s', avatar='%s', name='%s', viewed=%b, pinned=%b, startPosition=%d, slides=%s}",
				id, avatar, name, viewed, pinned, startPosition, slides);
	}
}

File: personalizatio-sdk/src/main/kotlin/com/personalizatio/stories/models/Story.kt
package com.personalizatio.stories.models

import org.json.JSONObject
import java.util.Objects

class Story(json: JSONObject) {
    val id: Int = json.optInt("id", 0)
    val avatar: String = json.optString("avatar", "")
    val name: String = json.optString("name", "")
    var isViewed: Boolean = json.optBoolean("viewed", false)
    val isPinned: Boolean = json.optBoolean("pinned", false)
    var startPosition: Int = json.optInt("start_position", 0)
    private val slides: MutableList<Slide> = ArrayList()

    init {
        val slidesJsonArray = json.optJSONArray("slides")
        if (slidesJsonArray != null) {
            for (i in 0 until slidesJsonArray.length()) {
                slides.add(Slide(slidesJsonArray.optJSONObject(i)))
            }
        }
    }

    fun getSlide(position: Int): Slide {
        return slides[position]
    }

    val slidesCount: Int
        get() = slides.size

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Story) return false
        return id == other.id
                && isViewed == other.isViewed
                && isPinned == other.isPinned
                && startPosition == other.startPosition
                && avatar == other.avatar
                && name == other.name
                && slides == other.slides
    }

    override fun hashCode(): Int {
        return Objects.hash(id, avatar, name, isViewed, isPinned, startPosition, slides)
    }

    override fun toString(): String {
        return "Story{id='$id', avatar='$avatar', name='$name', viewed=$isViewed, pinned=$isPinned, startPosition=$startPosition, slides=$slides}"
    }
}
